server/lib/pbench/process_tb.py

```diff
@@ -0,0 +1,131 @@
+import os
+import subprocess
+
+from collections import namedtuple
+from logging import Logger
+from pathlib import Path
+
+from pbench import PbenchConfig
+
+Results = namedtuple("Results", ["nstatus", "ntotal", "ntbs", "nerr"])
+
+
+class ProcessTb:
+    def __init__(self, config: PbenchConfig, logger: Logger):
+        """ Processes Tar ball, gets the tar ball from receiving directory
+        and copies it to the remote server by creating a subprocess which
+        call Agent's 'pbench-results-push' command.
+
+            Args -
+                config -- PbenchServer config object
+                logger -- logger object to use when emitting log entries during
+                            operation
+
+        """
+        self.config = config
+        self.logger = logger
+        self.token = self.config.get("pbench-server", "put-token")
+        if not self.token:
+            raise ValueError(
+                "No value for config option put-token in section pbench-server"
+            )
+        # receive_dir -- directory where tarballs are received from agent
+        self.receive_dir = self._get_receive_dir()
+
+    def _get_receive_dir(self) -> Path:
+        receive_dir_prefix = self.config.get(
+            "pbench-server", "pbench-receive-dir-prefix"
+        )
+        if not receive_dir_prefix:
+            raise ValueError(
+                "Failed: No value for config option pbench-receive-dir-prefix in section pbench-server"
+            )
+
+        receive_dir = Path(f"{receive_dir_prefix}-002").resolve()
+        if not receive_dir.is_dir():
+            raise NotADirectoryError(
+                "Failed: {} does not exist, or is not a directory", receive_dir
+            )
```

Comment:
I don't think this is the correct way to raise `NotADirectoryError`.  (For instance, my experimentation with it doesn't produce a nicely-formatted message.)

There are two signatures which `NotADirectoryError` inherits from [`OSError`](https://docs.python.org/3/library/exceptions.html#OSError):  one has a single argument (which I believe is intended to be a pre-formatted message) and one has several arguments, which get translated into a message when the exception is raised.

So, I think you should be using one of these two idioms:
```
            raise NotADirectoryError(
                f"Failed: {str(receive_dir)!r} does not exist, or is not a directory", 
            )
```
or
```
import errno
import os
...
            raise NotADirectoryError(
                errno.ENOTDIR,
                os.strerror(errno.ENOTDIR),
                str(receive_dir)
            )
```
I don't know whether you need the `str()` in the second one, but I'm not sure I like the second idiom much, and, if you go with the first one, it doesn't matter.

Reply:
I like the first one as well :)